File: apina-model/src/main/java/fi/evident/apina/model/Endpoint.java
package fi.evident.apina.model;

import fi.evident.apina.model.parameters.EndpointParameter;
import fi.evident.apina.model.type.ApiType;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static java.util.Collections.unmodifiableList;
import static java.util.Objects.requireNonNull;
import static java.util.stream.Collectors.joining;

/**
 * API endpoint reachable at an URL using given method and parameters. An example
 * of an endpoint is a single method on a controller.
 * Represents an single endpoint (e.g. a method in a controller)
 */
public final class Endpoint {

    /** Name of the original source element that specifies this endpoint */
    private final String name;

    /** URI template for the endpoint */
    private final URITemplate uriTemplate;

    private final List<EndpointParameter> parameters = new ArrayList<>();

    private final Optional<ApiType> responseBody;

    /** HTTP method for accessing the endpoint */
    private HTTPMethod method = HTTPMethod.GET;

    public Endpoint(String name, URITemplate uriTemplate, Optional<ApiType> responseBody) {
        this.name = requireNonNull(name);
        this.uriTemplate = requireNonNull(uriTemplate);
        this.responseBody = requireNonNull(responseBody);
    }

    public void addParameter(EndpointParameter parameter) {
        parameters.add(requireNonNull(parameter));
    }

    public List<EndpointParameter> getParameters() {
        return unmodifiableList(parameters);
    }

    public URITemplate getUriTemplate() {
        return uriTemplate;
    }

    public String getName() {
        return name;
    }

    public HTTPMethod getMethod() {
        return method;
    }

    public void setMethod(HTTPMethod method) {
        this.method = requireNonNull(method);
    }

    @Override
    public String toString() {
        return responseBody.map(ApiType::toString).orElse("void") + " " + name + parameters.stream().map(EndpointParameter::toString).collect(joining(", ", "(", ")")) + ": " + uriTemplate;
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/model/Endpoint.kt
package fi.evident.apina.model

import fi.evident.apina.model.parameters.EndpointParameter
import fi.evident.apina.model.parameters.EndpointPathVariableParameter
import fi.evident.apina.model.parameters.EndpointRequestBodyParameter
import fi.evident.apina.model.parameters.EndpointRequestParamParameter
import fi.evident.apina.model.settings.OptionalTypeMode
import fi.evident.apina.model.type.ApiType
import java.util.*

/**
 * API endpoint reachable at an URL using given method and parameters. An example
 * of an endpoint is a single method on a controller.
 * Represents an single endpoint (e.g. a method in a controller)
 */
class Endpoint(
    /** Name of the original source element that specifies this endpoint  */
    val name: String,

    /** URI template for the endpoint  */
    val uriTemplate: URITemplate,
    val responseBody: ApiType?,
    val generateUrlMethod: Boolean,
    val optionalTypeMode: OptionalTypeMode
) {

    private val _parameters = ArrayList<EndpointParameter>()

    /** HTTP method for accessing the endpoint  */
    var method = HTTPMethod.GET

    val parameters: List<EndpointParameter>
        get() = _parameters

    val urlParameters: List<EndpointParameter>
        get() = _parameters.filter { it !is EndpointRequestBodyParameter }

    fun addParameter(parameter: EndpointParameter) {
        _parameters += parameter
    }

    val requestBody: EndpointRequestBodyParameter?
        get() = _parameters.asSequence().filterIsInstance<EndpointRequestBodyParameter>().firstOrNull()

    val pathVariables: List<EndpointPathVariableParameter>
        get() = _parameters.filterIsInstance<EndpointPathVariableParameter>()

    val requestParameters: List<EndpointRequestParamParameter>
        get() = _parameters.filterIsInstance<EndpointRequestParamParameter>()

    override fun toString(): String = String.format("%s %s(%s): %s %s",
            responseBody?.toTypeScript(optionalTypeMode) ?: "void",
            name,
            _parameters.joinToString(", "),
            method,
            uriTemplate)
}
